src/prefect/environments/storage/s3.py

```diff
@@ -0,0 +1,168 @@
+import io
+import logging
+from typing import TYPE_CHECKING, Any, Dict
+
+
+import cloudpickle
+import pendulum
+from slugify import slugify
+
+import prefect
+from prefect.environments.storage import Storage
+
+if TYPE_CHECKING:
+    from prefect.core.flow import Flow
+
+
+class S3(Storage):
+    """
+    S3 storage class.  This class represents the Storage interface for Flows stored as
+    bytes in an S3 bucket.
+
+    This storage class optionally takes a `key` which will be the name of the Flow object
+    when stored in S3. If this key is not provided the Flow upload name will take the form
+    `slugified-flow-name/slugified-current-timestamp`.
+
+    Args:
+        - bucket (str): the name of the S3 Bucket to store Flows
+        - aws_access_key_id (str, optional): AWS access key id for connecting to S3.
+            Defaults to the value set in the environment variable
+            `AWS_ACCESS_KEY_ID` or `None`
+        - aws_secret_access_key (str, optional): AWS secret access key for connecting to S3.
+            Defaults to the value set in the environment variable
+            `AWS_SECRET_ACCESS_KEY` or `None`
+        - aws_session_token (str, optional): AWS session key for connecting to S3
+            Defaults to the value set in the environment variable
+            `AWS_SESSION_TOKEN` or `None`
+        - key (str, optional): a unique key to use for uploading a Flow to S3. This
+            is only useful when storing a single Flow using this storage object.
+    """
+
+    def __init__(
+        self,
+        bucket: str,
+        aws_access_key_id: str = None,
+        aws_secret_access_key: str = None,
+        aws_session_token: str = None,
+        key: str = None,
+    ) -> None:
+        self.flows = dict()  # type: Dict[str, str]
+        self._flows = dict()  # type: Dict[str, "prefect.core.flow.Flow"]
+        self.bucket = bucket
+        self.key = key
+
+        self.aws_access_key_id = aws_access_key_id
+        self.aws_secret_access_key = aws_secret_access_key
+        self.aws_session_token = aws_session_token
+
+        super().__init__()
+
+    def get_flow(self, flow_location: str) -> "Flow":
+        """
+        Given a flow_location within this Storage object, returns the underlying Flow (if possible).
+
+        Args:
+            - flow_location (str): the location of a flow within this Storage; in this case,
+                a file path where a Flow has been serialized to
+
+        Returns:
+            - Flow: the requested flow
+
+        Raises:
+            - ValueError: if the flow is not contained in this storage
+        """
+        if not flow_location in self.flows.values():
+            raise ValueError("Flow is not contained in this Storage")
+
+        stream = io.BytesIO()
+
+        logging.info("Downloading {} from {}".format(flow_location, self.bucket))
```

Comment:
Should we elevate this (and the other logs) to a prefect logger stored on the class?  Not a request, just curious if you think that would help with debugging.

Reply:
Yeah I think we should attach a prefect logger to the base storage class. Similar to how execution environments have it